util/tracing: segfault in `(*Span).Finish`   somewhere in sql/TestTrace

Found in CI in an unrelated PR: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_Test/2516403

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb4bca2]

goroutine 624262 [running]:
golang.org/x/net/trace.(*traceSet).Remove(0x0, 0xc004ac5860)
        /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:478 +0x22
golang.org/x/net/trace.(*trace).Finish(0xc004ac5860)
        /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:412 +0x13e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Finish(0xc00b0bc6c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:392 +0x112
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*TxnCoordSender).Send(0xc000d44d80, 0x500ebc0, 0xc003bff4d0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_coord_sender.go:528 +0x545
github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc0050eae70, 0x500ebc0, 0xc00df30e10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:744 +0x13c
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send(0xc0062f3dd0, 0x500ebc0, 0xc00df30e10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:919 +0x12e
github.com/cockroachdb/cockroach/pkg/kv.sendAndFill(0x500ebc0, 0xc00df30e10, 0xc0024d3fd8, 0xc005be1900, 0x0, 0x5)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:654 +0x107
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Run(0xc0062f3dd0, 0x500ebc0, 0xc00df30e10, 0xc005be1900, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:590 +0xe7
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Get(0xc0062f3dd0, 0x500ebc0, 0xc00df30e10, 0x408b8e0, 0xc0095b8e60, 0x8, 0xc0024d41c8, 0x14eeef3, 0xc0075ac460, 0xc0024d41c8, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:369 +0xbf
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).GetProtoTs(0xc0062f3dd0, 0x500ebc0, 0xc00df30e10, 0x408b8e0, 0xc0095b8e60, 0x5051b60, 0xc0041d56d0, 0x800000001, 0x220000c0024d4258, 0x1341378, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:391 +0x66
github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv.GetDescriptorByID(0x500ebc0, 0xc00df30e10, 0xc0062f3dd0, 0xc007f36e80, 0xc007f36e80, 0x0, 0xc, 0x4, 0x0, 0xc, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv/catalogkv.go:151 +0x191
github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv.GetAnyDescriptorByID(...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv/catalogkv.go:125
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getDescriptorFromStore(0xc00350f210, 0x500ebc0, 0xc00df30e10, 0xc0062f3dd0, 0xc007f36e80, 0xc007f36e80, 0x0, 0x1d00000001, 0x44c8f7f, 0x8, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:283 +0x125
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getObjectByName(0xc00350f210, 0x500ebc0, 0xc00df30e10, 0xc0062f3dd0, 0x44c8f78, 0x6, 0x436fb4f, 0x6, 0x44c8f7f, 0x8, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:432 +0x39c
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetTableByName(0xc00350f210, 0x500ebc0, 0xc00df30e10, 0xc0062f3dd0, 0x5053ae0, 0xc00845e680, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:453 +0x14b
github.com/cockroachdb/cockroach/pkg/sql/catalog/accessors.(*LogicalSchemaAccessor).GetObjectDesc(0xc0137b2480, 0x500ebc0, 0xc00df30e10, 0xc0062f3dd0, 0xc002fc7500, 0xc007f36e80, 0xc007f36e80, 0x0, 0x44c8f78, 0x6, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/accessors/logical_schema_accessors.go:152 +0x59b
github.com/cockroachdb/cockroach/pkg/sql.(*planner).LookupObject(0xc00350f3c0, 0x500ebc0, 0xc00df30e10, 0x0, 0x0, 0x0, 0x44c8f78, 0x6, 0x436fb4f, 0x6, ...)
...
```

It also repros readily with `make stress PKG=./pkg/sql TESTS=TestTrace`

cc @tbg  for triage.

--- a/pkg/util/tracing/span_test.go
+++ b/pkg/util/tracing/span_test.go
@@ -11,7 +11,6 @@
 package tracing
 
 import (
-	"context"
 	"regexp"
 	"strings"
 	"testing"
@@ -181,20 +180,4 @@ Span grandchild:
     === operation:grandchild _verbose:1
 `
 	require.Equal(t, exp, recToStrippedString(childRec))
-}
-
-func TestChildSpan(t *testing.T) {
-	tr := NewTracer()
-	// Set up non-recording span.
-	sp := tr.StartSpan("foo", WithForceRealSpan())
-	ctx := ContextWithSpan(context.Background(), sp)
-	// Since the parent span was not recording, we would expect the
-	// noop span back. However - we don't, we get our inputs instead.
-	// This is a performance optimization; there is a to-do in
-	// childSpan asking for its removal, but for now it's here to stay.
-	{
-		newCtx, newSp := ChildSpan(ctx, "foo")
-		require.Equal(t, ctx, newCtx)
-		require.Equal(t, sp, newSp)
-	}
 }

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -559,14 +559,13 @@ func childSpan(ctx context.Context, opName string, remote bool) (context.Context
 		opts.RemoteParent = sp.Meta()
 	}
 	newSpan := tr.startSpanGeneric(opName, opts)
-	if newSpan.isNoop() {
-		// Optimization: if we end up with a noop, return the inputs
-		// to avoid ContextWithSpan call below.
-		//
-		// TODO(tbg): this is unsound. We are returning the incoming
-		// context which may have a Span that could later start recording.
-		// So in effect that span may capture parts of two goroutines
-		// accidentally.
+	if sp.isNoop() && newSpan.isNoop() {
+		// Optimization: if we started and end up with a noop, we can return
+		// the original context to save on the ContextWithSpan alloc. Note
+		// that it is important that the incoming span was the noop span. If
+		// it was a real, non-recording span, it might later start recording.
+		// Besides, the caller expects to get their own span, and will
+		// .Finish() it, leading to an extra, premature call to Finish().
 		return ctx, sp
 	}
 	return ContextWithSpan(ctx, newSpan), newSpan

